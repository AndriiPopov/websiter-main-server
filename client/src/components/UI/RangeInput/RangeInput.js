import React from 'react';
import PropTypes from 'prop-types';

import classes from './RangeInput.module.css';

const RangeInput = props => {

    const handleChange = (e) => {
        let value = parseInt(e.target.value) || 0;
        if(props.min && value < props.min ) {
            value = props.min;
        }
        props.changed(value);
        hasBeenChanged = true;
    }

    let hasBeenChanged = false;
    const handleMouseDown = () => {

        hasBeenChanged = false;

        const dragStart = callbackStop => {
            const dragMouseUp = () => {
                window.removeEventListener('mouseup', dragMouseUp);
                callbackStop();
            }
            window.addEventListener('mouseup', dragMouseUp);
        }

        dragStart( () => {
            if (hasBeenChanged) {
                if(props.saveToHistory) {
                    props.saveToHistory();
                }
            }
        });
    };

    return (
        <div className={classes.Div}>
            {props.title}
            <input
                type="range"
                className={classes.Input}
                value={props.startValue}
                onChange={handleChange}
                step={props.step || 1}
                min={props.min || 0}
                max={props.max || 100}
                onMouseDown={handleMouseDown} />
        </div>
    );
};

RangeInput.propTypes = {
    sectionValues: PropTypes.object,
    boxesValues: PropTypes.object,
    min: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    max: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    step: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    isSection: PropTypes.bool,
    valueKey: PropTypes.string.isRequired,
    title: PropTypes.string,
    saveToHistory: PropTypes.func.isRequired,
    changed: PropTypes.func
};

export default RangeInput;