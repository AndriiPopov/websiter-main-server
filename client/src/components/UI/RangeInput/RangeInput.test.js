import React from 'react';
import {mount, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { RangeInput } from './RangeInput';

configure({adapter: new Adapter()});

describe('RangeInput', () => {
    it('should be defined', () => {
        expect(RangeInput).toBeDefined();
    });

    it('it should render button with onclick null when disabled', () => {
        const tree = mount(
            <RangeInput 
                step="1"
                min="0"
                max="100"
                sectionValues="{}"
                boxesValues="{}"
                changedStyleValues="[]" />
        );
        expect(toJson(tree)).toMatchSnapshot();
    });
});