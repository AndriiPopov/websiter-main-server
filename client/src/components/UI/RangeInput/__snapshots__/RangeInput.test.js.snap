// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeInput it should render button with onclick null when disabled 1`] = `
<RangeInput
  boxesValues="{}"
  changedStyleValues="[]"
  max="100"
  min="0"
  sectionValues="{}"
  step="1"
>
  <div
    className="Div"
  >
    <input
      className="Input"
      max="100"
      min="0"
      onChange={[Function]}
      onMouseDown={[Function]}
      step="1"
      type="range"
      value={0}
    />
  </div>
</RangeInput>
`;
